src/common-rest-api/src/main/java/org/niis/xroad/restapi/common/backup/service/BackupService.java

```diff
@@ -0,0 +1,138 @@
+import org.niis.xroad.restapi.exceptions.WarningDeviation;
+import org.niis.xroad.restapi.service.UnhandledWarningsException;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Service;
+
+import java.time.OffsetDateTime;
+import java.util.List;
+import java.util.Optional;
+
+import static org.niis.xroad.restapi.exceptions.DeviationCodes.WARNING_FILE_ALREADY_EXISTS;
+
+/**
+ * Backups service.
+ */
+@Slf4j
+@Service
+@PreAuthorize("isAuthenticated()")
+@RequiredArgsConstructor
+public class BackupService {
+    private final BackupRepository backupRepository;
+    private final BackupValidator backupValidator;
+    private final AuditDataHelper auditDataHelper;
+
+    /**
+     * Return a list of available backup files
+     *
+     * @return list of backup files
+     */
+    public List<BackupFile> getBackupFiles() {
+        return backupRepository.getBackupFiles();
+    }
+
+    /**
+     * Delete a backup file or throws an exception if the file does not exist
+     *
+     * @param filename backup file name to delete
+     * @throws BackupFileNotFoundException if backup file is not found
+     */
+    public void deleteBackup(String filename) throws BackupFileNotFoundException {
+        auditDataHelper.putBackupFilename(backupRepository.getAbsoluteBackupFilePath(filename));
+        if (getBackup(filename).isEmpty()) {
+            throw new BackupFileNotFoundException(filename);
+        }
+        backupRepository.deleteBackupFile(filename);
+    }
+
+    /**
+     * Read backup file's content
+     *
+     * @param filename backup file name to retrieve
+     * @return raw contents of file
+     * @throws BackupFileNotFoundException if backup file is not found
+     */
+    public byte[] readBackupFile(String filename) throws BackupFileNotFoundException {
+        if (getBackup(filename).isEmpty()) {
+            throw new BackupFileNotFoundException(filename);
+        }
+        return backupRepository.readBackupFile(filename);
+    }
+
+    /**
+     * Write uploaded backup file to disk. If ignoreWarnings=false, an exception is thrown when a file with
+     * the same name already exists. If ignoreWarnings=true, the existing file is overwritten.
+     *
+     * @param ignoreWarnings should the warning be ignored
+     * @param filename       backup file name
+     * @param fileBytes      file content
+     * @return newly created backup
+     * @throws InvalidFilenameException   if backup file's name is invalid and does not pass validation
+     * @throws UnhandledWarningsException if backup file with the same name already exists
+     *                                    and ignoreWarnings is false
+     * @throws BackupInvalidFileException if backup file is not a valid tar file or the first entry of the tar file
+     *                                    does not match to the first entry if the Security Server generated backup tar files
```

Comment:
"... first entry **of** the Security Server..."? Is this tar validation done currently? I couldn't find it with first glance, but in backupRepository.writeBackupFile there is a second backupValidator.isValidBackupFilename call, should that have been some sort of tar validation instead?

Reply:
upload functionality is out of scope of this story. This code was moved from security server and will be accomodated for central server in future stories